File: WdtcCore/src/main/java/org/wdt/wdtc/auth/yggdrasil/YggdrasilAccounts.java
package org.wdt.wdtc.auth.yggdrasil;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import lombok.Getter;
import org.apache.log4j.Logger;
import org.wdt.utils.gson.JSONObject;
import org.wdt.utils.gson.JSONUtils;
import org.wdt.utils.io.IOUtils;
import org.wdt.wdtc.auth.User;
import org.wdt.wdtc.auth.accounts.Accounts;
import org.wdt.wdtc.auth.accounts.AccountsInterface;
import org.wdt.wdtc.manger.FileManger;
import org.wdt.wdtc.manger.UrlManger;
import org.wdt.wdtc.utils.PlatformUtils;
import org.wdt.wdtc.utils.SkinUtils;
import org.wdt.wdtc.utils.WdtcLogger;

import java.io.IOException;
import java.io.PrintWriter;
import java.net.URL;
import java.net.URLConnection;

@Getter
public class YggdrasilAccounts implements AccountsInterface {
    private static final Logger logmaker = WdtcLogger.getLogger(YggdrasilAccounts.class);
    private final String url;
    private final String username;
    private final String password;

    public YggdrasilAccounts(String url, String username, String password) {
        this.password = password;
        this.username = username;
        this.url = url;
    }

    public String sendPostWithJson() throws IOException {
        URL requestUrl = new URL(url + "/api/yggdrasil/authserver/authenticate");
        String jsonStr = "{" +
                "\"username\":\"" + username + "\"," +
                "\"password\":\"" + password + "\"," +
                "\"requestUser\":true," +
                "\"agent\":{" +
                "\"name\":\"Minecraft\"," +
                "\"version\":1" +
                "}" +
                "}";
        URLConnection conn = requestUrl.openConnection();
        conn.setRequestProperty("content-type", "application/json");
        conn.setDoOutput(true);
        conn.setDoInput(true);
        PrintWriter out = new PrintWriter(conn.getOutputStream());
        out.print(jsonStr);
        out.flush();
        return IOUtils.toString(conn.getInputStream());
    }

    public UserInformation getUserInformation() throws IOException {
        return JSONObject.parseObject(sendPostWithJson(), UserInformation.class);
    }

    public YggdrasilTextures getYggdrasilTextures() {
        return new YggdrasilTextures(this);
    }

    @Override
    public User getUser() throws IOException {
        UserInformation UserInfo = getUserInformation();
        YggdrasilTextures textures = getYggdrasilTextures();
        User user = new User();
        user.setType(Accounts.AccountsType.Yggdrasil);
        JSONObject selectedProfile = new JSONObject(UserInfo.getSelectedProfile());
        user.setUserName(selectedProfile.getString("name"));
        user.setUuid(selectedProfile.getString("id"));
        user.setAccessToken(UserInfo.getAccessToken());
        user.setAPI(PlatformUtils.UrltoString(UrlManger.getLittleskinApi()));
        user.setAPIBase64(PlatformUtils.StringToBase64(user.getAPI()));
        SkinUtils utils = textures.getUtils();
        user.setHeadFile(utils.writeSkinHead());
        JSONUtils.ObjectToJsonFile(FileManger.getUsersJson(), user);
        logmaker.info(user);
        return user;
    }

    @Getter
    public static class UserInformation {
        public String accessToken;
        public String clientToken;
        public JsonArray availableProfiles;
        public JsonObject user;
        public JsonObject selectedProfile;
    }

}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/auth/accounts/yggdrasil/YggdrasilAccounts.kt
package org.wdt.wdtc.core.impl.auth.accounts.yggdrasil

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async
import okio.ByteString.Companion.encodeUtf8
import org.wdt.utils.gson.getString
import org.wdt.utils.gson.parseObject
import org.wdt.utils.gson.toJsonString
import org.wdt.utils.io.toStrings
import org.wdt.wdtc.core.openapi.auth.Accounts
import org.wdt.wdtc.core.openapi.auth.LoginUser
import org.wdt.wdtc.core.openapi.auth.User
import org.wdt.wdtc.core.openapi.manger.littleskinApiUrl
import org.wdt.wdtc.core.openapi.utils.ioAsync
import org.wdt.wdtc.core.openapi.utils.noNull
import org.wdt.wdtc.core.openapi.utils.runOnIO
import java.io.File
import java.io.PrintWriter
import java.net.URL

class YggdrasilAccounts(
	val url: String,
	private val userName: String,
	private val password: String
) : LoginUser {
	
	val textures = YggdrasilTextures(userName, url)
	
	private val sendPostWithJson = ioAsync {
		URL("$url/api/yggdrasil/authserver/authenticate").openConnection().apply {
			setRequestProperty("content-type", "application/json")
			doOutput = true
			doInput = true
		}.also {
			PrintWriter(it.outputStream).apply {
				print(PostJsonObject(userName, password).toJsonString())
				flush()
			}
		}.inputStream.use { input ->
			input.toStrings().parseObject<UserInformation>()
		}
	}
	
	private val headFile: Deferred<File> = ioAsync {
		textures.utils.writeSkinHead()
	}
	
	class PostJsonObject(
		@field:SerializedName("username") private val username: String,
		@field:SerializedName("password") private val password: String
	) {
		@SerializedName("requestUser")
		private val requestUser = true
		
		@SerializedName("agent")
		private val agent: Agent = Agent()
		
		class Agent(
			@field:SerializedName("name")
			private val name: String = "Minecraft",
			
			@field:SerializedName("version")
			private val version: Int = 1
		)
	}
	
	class UserInformation(
		var accessToken: String,
		var clientToken: String,
		var availableProfiles: JsonArray,
		var user: JsonObject,
		var selectedProfile: JsonObject,
	)
	
	override suspend fun getUser(): User = runOnIO {
		val metaDataDeferred = async { littleskinApiUrl.toStrings() }
		sendPostWithJson.await().let {
			User(
				userName, it.accessToken, Accounts.AccountsType.YGGDRASIL, it.selectedProfile.getString("id"), headFile.await()
			).apply {
				this.metaData = metaDataDeferred.await()
				this.base64Data = metaData.noNull().encodeUtf8().base64()
			}
		}
	}
}
